### PR TITLE
Fix/cap slippage comp

### LINKED ISSUE #1581: Capping slippage component of margin
At the moment it seems that an order book with unreliable buys / sells can lead to undesirable closeouts getting triggered (@Jiajia-Cui is confirming with a feature test which we can review and then update to check we really have a fix). 

The reason is the slippage component of the margin (see also https://github.com/vegaprotocol/specs/issues/1567). 

Imagine a party A that's short 1 ETHUSD. If the buys sells are 
```
buy 1 @ 15000 
buy 10 @ 14900 
and
sell 1 @ 16000
sell 10 @ 16100 
```

then, if the last mark price was `15900` then the slippage charge is `16000 - 15900 = 100` (because to close the party out I need to execute a network trade on behalf of the distressed party, that would take its position to `0` i.e. the network needs to buy `1` so I need to compare to the price for selling `1` unit. The RF part is probably something like `15900 x 0.1 = 1590` so the slippage part is modest component of the maintenance margin. So far so good (or at least no disaster). 

Now imagine the book becomes: 
```
buy 1 @ 15000 
buy 10 @ 14900 
and
sell 1 @ 100 000
sell 10 @ 100 100 
```

The slippage is now `100 000 - 15900 = 84 100` which is a *lot* and will most likely initiate a closeout, draining the party's general and margin accounts (and bond if they happen to be an LP). 

So far this is the problem description. 

## Proposed solution 

I think we'd want both. 

### Cap the slippage component

Introduce a new optional market parameter `market.maxSlippageFraction` with default of `0.1` i.e. `10%`. The slippage component will never be allowed to be more that `min(slippage from book, market.maxSlippageFraction x size x mark price)`. 

The spec change PR for this is: https://github.com/vegaprotocol/specs/pull/1583

### Do not execute closeout trades that would reach volume too far out

Here, too far out would be IMO: 
1. volume outside price monitoring bounds 
1. volume outside LP liquidity range `market.liquidity.priceRange`.

### PR DESCRIPTION
**Pull request by davidsiska-vega**
This deals with the first suggestion in https://github.com/vegaprotocol/specs/issues/1581 

close #1581 .

### CODE REVIEW

**Comment by witgaw** [comment]
It looks good, the only worry I have is that the larger the position the more likely we are to switch to a linear, rather than superlinear, slippage component of the margin level which may encourage build up of very large positions relative to the market.

File: protocol/0019-MCAL-margin_calculator.md [hunk 1]
```diff
@@ -2,16 +2,41 @@
...
-8. A feature test that checks margin in case market PDP > 0 is created and passes. (<a name="0019-MCAL-008" href="#0019-MCAL-008">0019-MCAL-008</a>)
-9. For each market and each party which has either orders or positions on the market, the API provides the 4 margin levels.  (<a name="0019-MCAL-009" href="#0019-MCAL-009">0019-MCAL-009</a>)
-10. A feature test that checks margin in case market PDP < 0 is created and passes. (<a name="0019-MCAL-010" href="#0019-MCAL-010">0019-MCAL-010</a>)
+- Get four margin levels for one or more parties (<a name="0019-MCAL-001" href="#0019-MCAL-001">0019-MCAL-001</a>)
+
+- Margin levels are correctly calculated against riskiest long and short positions (<a name="0019-MCAL-002" href="#0019-MCAL-002">0019-MCAL-002</a>)
+
+- Zero position and zero orders results in all zero margin levels (<a name="0019-MCAL-003" href="#0019-MCAL-003">0019-MCAL-003</a>)
+
+- If `riskiest long > 0` and there are no bids on the order book, the `exit price` is equal to the initial mark price, as set by a market parameter. (<a name="0019-MCAL-004" href="#0019-MCAL-004">0019-MCAL-004</a>)
```
Comment (witgaw) [inline protocol/0019-MCAL-margin_calculator.md:11]: Wait, I know it's not your change (just the formatting), but I think this is dated, we no longer have that market parameter...

Comment (davidsiska-vega) [reply protocol/0019-MCAL-margin_calculator.md:11]: Good point, let's get rid of that!

**Comment by davidsiska-vega** [comment]
> It looks good, the only worry I have is that the larger the position the more likely we are to switch to a linear, rather than superlinear, slippage component of the margin level which may encourage build up of very large positions relative to the market.

Good point, I made the additional factor 2d so that one entry multiplies size and the other size^2.

File: protocol/0019-MCAL-margin_calculator.md [hunk 2]
```diff
@@ -2,16 +2,39 @@
...
+
+- If `riskiest short < 0 && 0 < *sum of absolute volume of order book offers* < riskiest short`, the `exit price` is equal to the *volume weighted price of the order book offers*.  (<a name="0019-MCAL-007" href="#0019-MCAL-007">0019-MCAL-007</a>)
+
+- A feature test that checks margin in case market PDP > 0 is created and passes. (<a name="0019-MCAL-008" href="#0019-MCAL-008">0019-MCAL-008</a>)
+
+- For each market and each party which has either orders or positions on the market, the API provides the 4 margin levels.  (<a name="0019-MCAL-009" href="#0019-MCAL-009">0019-MCAL-009</a>)
+
+- A feature test that checks margin in case market PDP < 0 is created and passes. (<a name="0019-MCAL-010" href="#0019-MCAL-010">0019-MCAL-010</a>)
+
+- If a party is short `1` unit and the mark price is `15 900` and `market.maxSlippageFraction[1] = 0.25`, `market.maxSlippageFraction[2] = 0.25` and `RF shourt = 0.1` and order book is
```
Comment (TomMcL) [inline protocol/0019-MCAL-margin_calculator.md]:
```suggestion
- If a party is short `1` unit and the mark price is `15 900` and `market.maxSlippageFraction[1] = 0.25`, `market.maxSlippageFraction[2] = 0.25` and `RF short = 0.1` and order book is
```

File: protocol/0019-MCAL-margin_calculator.md [hunk 3]
```diff
@@ -2,16 +2,39 @@
...
+
+```book
```
Comment (gordsport) [inline protocol/0019-MCAL-margin_calculator.md:25]:
```suggestion
    ```book
```

File: protocol/0019-MCAL-margin_calculator.md [hunk 4]
```diff
@@ -2,16 +2,39 @@
...
+buy 1 @ 15 000
```
Comment (gordsport) [inline protocol/0019-MCAL-margin_calculator.md:26]:
```suggestion
    buy 1 @ 15 000 
```

File: protocol/0019-MCAL-margin_calculator.md [hunk 5]
```diff
@@ -2,16 +2,39 @@
...
+
+- For each market and each party which has either orders or positions on the market, the API provides the 4 margin levels.  (<a name="0019-MCAL-009" href="#0019-MCAL-009">0019-MCAL-009</a>)
+
+- A feature test that checks margin in case market PDP < 0 is created and passes. (<a name="0019-MCAL-010" href="#0019-MCAL-010">0019-MCAL-010</a>)
+
+- If a party is short `1` unit and the mark price is `15 900` and `market.maxSlippageFraction[1] = 0.25`, `market.maxSlippageFraction[2] = 0.25` and `RF shourt = 0.1` and order book is
+
+```book
+buy 1 @ 15 000 
+buy 10 @ 14 900
```
Comment (gordsport) [inline protocol/0019-MCAL-margin_calculator.md:27]:
```suggestion
    buy 10 @ 14 900 
```

File: protocol/0019-MCAL-margin_calculator.md [hunk 6]
```diff
@@ -2,16 +2,39 @@
...
+- For each market and each party which has either orders or positions on the market, the API provides the 4 margin levels.  (<a name="0019-MCAL-009" href="#0019-MCAL-009">0019-MCAL-009</a>)
+
+- A feature test that checks margin in case market PDP < 0 is created and passes. (<a name="0019-MCAL-010" href="#0019-MCAL-010">0019-MCAL-010</a>)
+
+- If a party is short `1` unit and the mark price is `15 900` and `market.maxSlippageFraction[1] = 0.25`, `market.maxSlippageFraction[2] = 0.25` and `RF shourt = 0.1` and order book is
+
+```book
+buy 1 @ 15 000 
+buy 10 @ 14 900 
+and
```
Comment (gordsport) [inline protocol/0019-MCAL-margin_calculator.md:28]:
```suggestion
    and
```

**Review by gordsport — DISMISSED** [review]
Hey @davidsiska-vega 
Ive added suggestions that will align the list formatting as needed

**Review by TomMcL — APPROVED** [review]

**Review by Jiajia-Cui — APPROVED** [review]
LGTM

**Review by witgaw — APPROVED** [review]

**Pull request merged**